File: util/x58/scripts/x58_cmos_probe.py
#!/usr/bin/env python3
"""Read the MSI X58 early-policy CMOS inputs through /dev/port.

Only the two CMOS selector ports are written.  Their original values are
restored in a finally block; neither CMOS data port is ever written.  The
ICH10 upper-bank enable is checked read-only before any extended access.
"""

from __future__ import annotations

import os
import struct


ICH10_LPC_CONFIG = "/sys/bus/pci/devices/0000:00:1f.0/config"
ICH10_RCBA_REGISTER = 0xF0
ICH10_RC_OFFSET = 0x3400
ICH10_RC_U128E = 1 << 2


STANDARD_ADDRESSES = (
	0x0E,
	0x57,
	0x58,
	0x5E,
	0x5F,
	0x69,
	0x70,
	0x71,
	0x72,
	0x73,
	0x74,
	0x75,
)
EXTENDED_ADDRESSES = (
	0x07,
	0x0A,
	0x0B,
	0x0C,
	0x0D,
	0x11,
	0x12,
	0x1E,
	0x28,
	0x44,
	0x80,
	0x81,
	0x82,
	0x88,
	0x89,
	0x8E,
	0xA8,
	0xC6,
	0xCA,
	0xDB,
	0xDD,
	0xF1,
	0xF5,
	0xF6,
	0xF7,
	0xF8,
	0xF9,
)


def read_port(fd: int, port: int) -> int:
	value = os.pread(fd, 1, port)
	if len(value) != 1:
		raise OSError(f"short /dev/port read at 0x{port:x}")
	return value[0]


def write_port(fd: int, port: int, value: int) -> None:
	if os.pwrite(fd, bytes((value & 0xFF,)), port) != 1:
		raise OSError(f"short /dev/port write at 0x{port:x}")


def read_ich10_rtc_configuration() -> tuple[int, int]:
	with open(ICH10_LPC_CONFIG, "rb", buffering=0) as config:
		config.seek(ICH10_RCBA_REGISTER)
		raw_rcba = config.read(4)
	if len(raw_rcba) != 4:
		raise OSError("short ICH10 RCBA PCI-config read")
	rcba = struct.unpack("<I", raw_rcba)[0]
	if not rcba & 1:
		raise RuntimeError(f"ICH10 RCBA decode is disabled: 0x{rcba:08x}")
	rc_address = (rcba & 0xFFFFC000) + ICH10_RC_OFFSET
	mem_fd = os.open("/dev/mem", os.O_RDONLY | os.O_CLOEXEC)
	try:
		raw_rc = os.pread(mem_fd, 4, rc_address)
	finally:
		os.close(mem_fd)
	if len(raw_rc) != 4:
		raise OSError(f"short /dev/mem read at 0x{rc_address:x}")
	return rcba, struct.unpack("<I", raw_rc)[0]


def main() -> None:
	rcba, rtc_configuration = read_ich10_rtc_configuration()
	if not rtc_configuration & ICH10_RC_U128E:
		raise SystemExit(
			f"REFUSED: RCBA=0x{rcba:08x} RC=0x{rtc_configuration:08x}; "
			"upper-128 RTC decode is disabled and 0x72/0x73 would alias 0x70/0x71"
		)
	fd = os.open("/dev/port", os.O_RDWR | os.O_CLOEXEC)
	try:
		saved_index_70 = read_port(fd, 0x70)
		saved_index_72 = read_port(fd, 0x72)
		try:
			alt_gp_smi_en_low = read_port(fd, 0x538)
			standard: dict[int, int] = {}
			for address in STANDARD_ADDRESSES:
				write_port(fd, 0x70, address | 0x80)
				read_port(fd, 0x61)
				standard[address] = read_port(fd, 0x71)

			extended: dict[int, int] = {}
			for address in EXTENDED_ADDRESSES:
				write_port(fd, 0x72, address & 0x7F)
				extended[address] = read_port(fd, 0x73)
		finally:
			write_port(fd, 0x72, saved_index_72)
			write_port(fd, 0x70, saved_index_70)
	finally:
		os.close(fd)

	print(
		f"RCBA={rcba:08x} RTC_RC={rtc_configuration:08x} U128E=1 "
		f"INDEX70={saved_index_70:02x} INDEX72={saved_index_72:02x} "
		f"ALT_GP_SMI_EN_LOW={alt_gp_smi_en_low:02x} "
		f"FIELD_6_4={(alt_gp_smi_en_low >> 4) & 7}"
	)
	for address in STANDARD_ADDRESSES:
		label = "DIAG_0E" if address == 0x0E else f"STD_{address:02X}"
		print(f"CMOS_{label}={standard[address]:02x}")
	for address in EXTENDED_ADDRESSES:
		print(f"CMOS_EXT_{address:02X}={extended[address]:02x}")


if __name__ == "__main__":
	main()

File: util/x58/scripts/x58_cmos_probe.c
/* SPDX-License-Identifier: GPL-2.0-only */

#define _GNU_SOURCE
#include <fcntl.h>
#include <errno.h>
#include <stdint.h>
#include <stdio.h>
#include <string.h>
#include <sys/io.h>
#include <unistd.h>

#define ICH10_LPC_CONFIG "/sys/bus/pci/devices/0000:00:1f.0/config"
#define ICH10_RCBA_REGISTER 0xf0
#define ICH10_RC_OFFSET 0x3400
#define ICH10_RC_U128E (1U << 2)

static unsigned char read_extended_cmos(unsigned char address)
{
	/* Port 0x72 selects bits 6:0 inside the enabled upper 128-byte bank. */
	outb(address & 0x7f, 0x72);
	return inb(0x73);
}

static int read_u32_at(const char *path, off_t offset, uint32_t *value)
{
	int fd = open(path, O_RDONLY | O_CLOEXEC);
	ssize_t count;

	if (fd < 0)
		return -1;
	count = pread(fd, value, sizeof(*value), offset);
	if (close(fd) != 0 || count != sizeof(*value))
		return -1;
	return 0;
}

static unsigned char read_standard_cmos(unsigned char selector)
{
	/* Bit 7 keeps NMI disabled exactly as the MSI early wrapper does. */
	outb(selector | 0x80, 0x70);
	(void)inb(0x61);
	return inb(0x71);
}

int main(void)
{
	static const unsigned char addresses[] = {
		0x81, 0x82, 0x88, 0x89, 0x8e, 0xca, 0xf1, 0xf5,
	};
	unsigned char saved_index_70;
	unsigned char saved_index_72;
	unsigned char alt_gp_smi_en_low;
	uint32_t rcba;
	uint32_t rtc_configuration;
	size_t i;

	if (read_u32_at(ICH10_LPC_CONFIG, ICH10_RCBA_REGISTER, &rcba) != 0) {
		fprintf(stderr, "ICH10 RCBA read: %s\n", strerror(errno));
		return 1;
	}
	if (!(rcba & 1) ||
	    read_u32_at("/dev/mem", (rcba & 0xffffc000U) + ICH10_RC_OFFSET,
		&rtc_configuration) != 0) {
		fprintf(stderr, "ICH10 RTC configuration read failed (RCBA=%08x): %s\n",
		       rcba, strerror(errno));
		return 1;
	}
	if (!(rtc_configuration & ICH10_RC_U128E)) {
		fprintf(stderr,
			"REFUSED: RCBA=%08x RTC_RC=%08x; upper-128 RTC decode is disabled and 72/73 would alias 70/71\n",
			rcba, rtc_configuration);
		return 1;
	}

	if (ioperm(0x70, 4, 1) != 0 || ioperm(0x538, 1, 1) != 0) {
		fprintf(stderr, "ioperm: %s\n", strerror(errno));
		return 1;
	}

	saved_index_70 = inb(0x70);
	saved_index_72 = inb(0x72);
	/* Match the original wrapper's byte read; only bits 6:4 are consumed. */
	alt_gp_smi_en_low = inb(0x538);
	printf("RCBA=%08x RTC_RC=%08x U128E=1 INDEX70=%02x INDEX72=%02x ALT_GP_SMI_EN_LOW=%02x FIELD_6_4=%u\n",
	       rcba, rtc_configuration, saved_index_70, saved_index_72,
	       alt_gp_smi_en_low,
	       (alt_gp_smi_en_low >> 4) & 7);
	printf("CMOS_DIAG_0E=%02x\n", read_standard_cmos(0x0e));
	for (i = 0; i < sizeof(addresses); i++)
		printf("CMOS_%02X=%02x\n", addresses[i],
		       read_extended_cmos(addresses[i]));

	/* Restore both selectors. No CMOS data register is ever written. */
	outb(saved_index_72, 0x72);
	outb(saved_index_70, 0x70);
	return 0;
}
